File: app/src/main/java/com/example/shibegallery/repository/remote/RetrofitInstance.java
package com.example.shibegallery.repository.remote;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

public class RetrofitInstance {

    private RetrofitInstance() {

    }

    private static class RetrofitHolder {
        private static final Retrofit retrofit = new Retrofit
                .Builder()
                .baseUrl("http://shibe.online/api/")
                .addConverterFactory(MoshiConverterFactory.create())
                .client(getClient())
                .build();

        private static OkHttpClient getClient() {
            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BASIC);
            return new OkHttpClient.Builder().addInterceptor(interceptor).build();
        }
    }

    public static ShibeService getShibeService() {
        return RetrofitHolder.retrofit.create(ShibeService.class);
    }
}

File: appkotlin/src/main/java/com/example/appkotlin/repository/remote/RetrofitInstance.kt
package com.example.appkotlin.repository.remote

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

object RetrofitInstance {

    // JAVAish Kotlin
    /*private fun getClient() : OkHttpClient {
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.level = HttpLoggingInterceptor.Level.BASIC
        return OkHttpClient.Builder().addInterceptor(loggingInterceptor).build()
    }*/

    private val client by lazy {
        HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BASIC }
                .let { OkHttpClient.Builder().addInterceptor(it).build() }
    }

    private val retrofit by lazy {
        Retrofit.Builder()
                .baseUrl("http://shibe.online/api/")
                .addConverterFactory(MoshiConverterFactory.create())
                .client(client)
                .build()
    }

    val shibeService = retrofit.create(ShibeService::class.java)
}